### ISSUE TITLE
generic class with single constructor input can be inferred as a distributed union

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#1217]
### Summary

```py
class A[T]:
    def __init__(self, value: T):
        self._value = value

def f(value: int | str):
    _: A[int] | A[str] = A(value)  # Object of type `A[int | str]` is not assignable to `A[int] | A[str]`
```

the value here could be safely inferred as `A[int] | A[str]`, as there is only one input value

**Comment by sharkdp** [comment]
Thank you for reporting this.

> the value here could be safely inferred as `A[int] | A[str]`, as there is only one input value

It seems to me like there are additional requirements that need to be fulfilled in order for that property to hold (not just the "one input value" constraint)? It might work for `int` and `str`, as there is an instance lay-out conflict, but for general classes, I don't think it's true that `A(value)` with `value: X | Y` should be assignable to `A[X] | A[Y]`, because there could be subclasses that derive from `X` and `Y`, and `value` might be an instance of those subclasses.

It doesn't seem like any other type checker allows your snippet to pass without errors. Do you have a concrete real-world use case where you see this problem?

**Comment by KotlinIsland** [comment]
i did run into this at one point, but i can't remember the specifics, i think it was creating a list

in the case of covariant types `A[int] | A[str]` is a subtype of `A[int | str]`. so we would be inferring the more specific type

**Comment by KotlinIsland** [comment]
> It seems to me like there are additional requirements that need to be fulfilled in order for that property to hold

yeah, it's more complicated than i initially thought, and now i think not worth it
